doc/OVERVIEW.md

```diff
@@ -0,0 +1,484 @@
+
+As a result, **Three** [`PluginContext`](#contexts) will be created per Wasm VM and each of them corresponds to each of the above filter configurations (the top `configuration` fields at 18000, 18001, and 18002 respectively).
+
+See [example.yaml](../examples/shared_queue/envoy.yaml) for a full example.
+
+# Proxy-Wasm Go SDK API
+
+So far we have explaind the concepts and *plugin configs*. Now we are ready to dive into the API of this SDK.
+
+## *Contexts*
+
+*Contexts* are collection of interfaces in Proxy-Wasm Go SDK, and all of them are mapped to the concepts  explained above. They are defined in [types](../proxywasm/types) package, and developers are supposed to implement these interfaces in order to extend network proxies.
+
+There are four types of contexts: `VMContext`, `PluginContext`, `TcpContext` and `HttpContext`. Their relationship and how they are mapped to the concepts above can be described as the following diagram:
+
+```
+                    Wasm Virtual Machine
+                      (.vm_config.code)
+┌────────────────────────────────────────────────────────────────┐
+│  Your program (.vm_config.code)                TcpContext      │
+│          │                                  ╱ (Tcp stream)     │
+│          │ 1: 1                            ╱                   │
+│          │         1: N                   ╱ 1: N               │
+│      VMContext  ──────────  PluginContext                      │
+│                                (Plugin)   ╲ 1: N               │
+│                                            ╲                   │
+│                                             ╲  HttpContext     │
+│                                               (Http stream)    │
+└────────────────────────────────────────────────────────────────┘
+```
+
+To summarize,
+
+1) `VMContext` corresponds to each `.vm_config.code`, and only one `VMContext` exists in each VM.
+2) `VMContext` is the parent of *PluginContexts*, and is responsible for creating arbitrary number of `PluginContexts`.
+3) `PluginContext` corresponds to a *Plugin* instance. That means, a `PluginContext` corresponds to a *Http Filter* or *Network Filter* or maybe *Wasm Service*, configured via `.configuration` field in the *plugin config*.
+4) `PluginContext` is the parent of `TcpContext` and `HttpContexts`, and is responsible for creating arbitrary number of these contexts when it is configured at *Http Filter* or *Network Filter*.
+5) `TcpContext` is responsible for handling each Tcp stream.
+6) `HttpContexts` is responsible for handling each Http stream.
+
+So all you have to do is implement `VMContext` and `PluginContext`. And if you want to plug in to *Http Filter* or *Network Filter*, then implement `HttpContext` or *TcpFilter* respectively.
```

Comment:
Are you perhaps intending `TcpContext` instead of *TcpFilter*?

Reply:
Thanks!